ITISFoundation/osparc-simcore
# Is1884/service integration library

## What do these changes do?

Creating a service integration library in osparc to simplify submission and integration of services in the platform. 

This library includes now  most the functionality currently in the [cookiecutter-osparc-service](https://github.com/ITISFoundation/cookiecutter-osparc-service) regarding:
- tooling ( e.g. inject metadata in images, )
- testing fixtures/assert helpers
- versioning

**NOTE**: A PR in  [cookiecutter-osparc-service](https://github.com/ITISFoundation/cookiecutter-osparc-service) including this library will follow to this.


### ``simcore-service-integrator`` entrypoint

Commands of ``simcore-service-integrator`` CLI:
```cmd
$ simcore-service-integrator  --help
Usage: simcore-service-integrator [OPTIONS] COMMAND [ARGS]...

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
  bump-version           Bumps target version in metadata
  get-version            Prints to output requested version
  run-creator            Creates a sh script that uses jq tool to retrieve...
  update-compose-labels  Update a docker-compose file with json files in a...
```


#### tooling


A replacement for the old Makefile recipes might be:

```Makefile
service.cli/run: $(METADATA)
	# Updates adapter script from metadata in $<
	simcore-service-integrator run-creator --metadata $< --runscript $@

docker-compose-meta.yml: $(METADATA)
	# Injects metadata from $< as labels
	simcore-service-integrator update-compose-labels --compose $@ --metadata $<

```
#### testing plugin

Created a pytest-plugin from submodule ``service_integration.pytest_plugin`` with fixtures and helper assert function.

A sample of ``conftest.py`` in target repo

```python
import pytest

pytest_plugins = [
    "service_integration.pytest_plugin.folder_structure",
    "service_integration.pytest_plugin.validation_data",
]

current_dir = Path(sys.argv[0] if __name__ == "__main__" else __file__).resolve().parent


@pytest.fixture(scope='session')
def project_slug_dir() -> Path:
    project_slug_dir = current_dir.parent
    assert project_slug_dir.exists()
    return project_slug_dir

```

#### versioning

The publication of a service requires different type of versions explictly set in the ``metadata`` file or deduced by this tool. These versions are

- Service ``version``: author's version of the service. This is how is shown to the user e.g. ``3.2.1`` or ``matterhorn``
- Service ``integration_version``: version of the integration interface

Tools to increase versions by specifying ``patch, minor, major``.
```cmd
$ simcore-service-integrator bump-version --help
Usage: simcore-service-integrator bump-version [OPTIONS] [[integration-
                                               version|version]]

  Bumps target version in metadata

Options:
  --upgrade [major|minor|patch]  [required]
  --metadata-file PATH           The metadata yaml file
  --help                         Show this message and exit.
```

so a replacement Makefile recipes might be

```Makefile

.PHONY: version-service-patch version-service-minor version-service-major
version-service-patch version-service-minor version-service-major: $(METADATA) ## kernel/service versioning as patch
	simcore-service-integrator bump-version --metadata-file $<  --upgrade $(subst version-service-,,$@)

.PHONY: version-integration-patch version-integration-minor version-integration-major
version-integration-patch version-integration-minor version-integration-major: $(METADATA) ## integration versioning as patch (bug fixes not affecting API/handling), minor/major (backwards-compatible/INcompatible API changes)
	simcore-service-integrator bump-version --metadata-file $<  --upgrade $(subst version-integration-,,$@) integration-version


CURRENT_VERSION := $(shell simcore-service-integrator get-version --metadata-file $(METADATA))
CURRENT_INTEGRATION_VERSION := $(shell simcore-service-integrator get-version --metadata-file $(METADATA) integration-version)

```


## Related issue number


#1884 

## How to test


```cmd
make devenv
cd packages/service-integration
make install-dev
make tests
```

To try in a real project, one can install the from repo as

```cmd
pip install git+https://github.com/pcrespov/osparc-simcore.git@is1884/integration-library#egg=simcore-service-integration&subdirectory=packages/service-integration
```
See migration example in PR ITISFoundation/osparc-python-runner#6

## Checklist
